### PR TITLE
Add support for inline script and datums

### LINKED ISSUE #681: Vasil features: add support for TransactionOutput inline scripts and datums
Follow these changes in CSL: https://github.com/Emurgo/cardano-serialization-lib/pull/450

Modify our (de)serialization and domain types accordingly

Part of https://github.com/Plutonomicon/cardano-transaction-lib/issues/567

### PR DESCRIPTION
**Pull request by klntsky**
Closes #681

### Pre-review checklist

- [x] All code has been formatted using our config (`make format` for Purescript, `nixpkgs-fmt` for Nix)
- [x] All Purescript imports are explicit, including constructors
- [ ] **All** existing examples have been run locally against a fully-synced testnet node
- [x] The integration and unit tests have been run (`npm run test`) locally
- [ ] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`)

### CODE REVIEW

File: src/Deserialization/UnspentOutput.purs [hunk 1]
```diff
@@ -80,7 +89,15 @@ convertOutput output = do
     dataHash =
       getDataHash maybeFfiHelper output <#>
         asOneOf >>> toBytes >>> T.DataHash
-  pure $ T.TransactionOutput { address, amount, dataHash }
+  datum <- getPlutusData maybeFfiHelper output # traverse convertPlutusData
```
Comment (errfrom) [inline src/Deserialization/UnspentOutput.purs:92]: I would suggest to add an auxiliary function `convertOutputDatum` and utilize the CSL `.has_data_hash` and `.has_plutus_data` functions to explicitly show that we are working with alternative values

File: src/Cardano/Types/Transaction.purs [hunk 2]
```diff
@@ -768,24 +769,12 @@ instance Monoid AuxiliaryData where
...
-derive instance Generic NativeScript _
-
-instance Show NativeScript where
-  show x = genericShow x
-
 newtype TransactionOutput = TransactionOutput
   { address :: Address
   , amount :: Value
   , dataHash :: Maybe DataHash
+  , datum :: Maybe PlutusData
```
[the remark this answers is not in the document]
Comment (errfrom) [reply src/Cardano/Types/Transaction.purs:776]: @klntsky I noticed this is exactly the way our `Plutus.TransactionOutput` is defined. Is there any particular reason why we might not want to use `OutputDatum` in `Cardano. TransactionOutput`?

[the remark this answers is not in the document]
Comment (errfrom) [reply src/Cardano/Types/Transaction.purs:776]: @klntsky From the CSL release notes: 
> the data-hash and the plutus-data are alternative values and can NOT be both present in an output both at the same time.

[the remark this answers is not in the document]
Comment (klntsky) [reply src/Cardano/Types/Transaction.purs:776]: Sure, then I'll use `TransactionOutput` in `Cardano.`

File: src/Deserialization/UnspentOutput.purs [hunk 3]
```diff
@@ -122,6 +148,19 @@ foreign import getOutput :: TransactionUnspentOutput -> TransactionOutput
...
 foreign import getTransactionIndex :: TransactionInput -> Int
 foreign import getAddress :: TransactionOutput -> Address
+foreign import getPlutusData
+  :: MaybeFfiHelper -> TransactionOutput -> Maybe PlutusData
+
+foreign import getScriptRef
+  :: MaybeFfiHelper -> TransactionOutput -> Maybe ScriptRef
+
+foreign import withScriptRef
+  :: forall a
```
Comment (ngua) [inline src/Deserialization/UnspentOutput.purs:158]:
```suggestion
  :: forall (a :: Type)
```

**Review by ngua — APPROVED** [review]
Looks good, a few minor comments

File: src/Types/ScriptLookups.purs [hunk 4]
```diff
@@ -1016,13 +1036,17 @@ processConstraint mpsMap osMap = do
       let amount = fromPlutusValue plutusValue
       runExceptT do
         -- Don't write `let dataHash = datumHash datum`, see [datumHash Note]
-        dataHash <- except $ note (CannotHashDatum dat)
+        _dataHash <- except $ note (CannotHashDatum dat)
```
Comment (vvtran) [inline src/Types/ScriptLookups.purs:1039]: Could use `void` although see below

**Review by vvtran — APPROVED** [review]

**Pull request merged**